### ISSUE TITLE
Work out how much extracting XML in attachments improves performance

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#3058]
XML forms are now being stored in attachments instead of in the `content` property of the report's json object: #3009 

I also wrote a migration script that does this to all existing forms. It is currently in a branch. Here is the commit: https://github.com/medic/medic-api/commit/7ae514f6596985577203880774080ab16a6a0649

This migration script is very disruptive, causing CHWs to re-download all reports.

However, it may significantly improve view generation (and other misc) performance, especially on older phones.

We should:
 - Devise and execute on a way to show performance differences between all your reports being the old style or the new style, e.g. speed / memory usage of generating all views on a slow device
 - Consider performance given that existing CHP's views are already generated, e.g. consider general app memory usage, smaller scale things
 - Decide whether to
 -- Throw the migration away
 -- Merge the migration and take the download penalty
 -- Convert the migration to a standalone script and have this in our back pocket to strategically run this on certain systems, or even on certain CHP's reports

**Comment by sglangevin** [comment]
One other comment re: the migration. If we run the migration on older docs, we need to make sure that we also extract geo data from the old docs and put that in JSON. We have reports that were sent in as late as early November last year that do not have geo data exposed in JSON. This is necessary for analytics (cc @michaelkohn). We also need to review to see if there are any other fields that need to be exposed in JSON during this migration.

**Comment by SCdF** [comment]
Correct me if I'm wrong @michaelkohn but I don't think you're actually doing that anywhere. It was talked about but never implemented.

**Comment by michaelkohn** [comment]
Doing what?

**Comment by SCdF** [comment]
@michaelkohn read @sglangevin's comment: I don't think we're currently using geo data only available in XML, we've just talked about it.

**Comment by michaelkohn** [comment]
@scdf we are definitely using geo data and in many cases it is only available in the XML so if you are going to get rid of the XML, I need that GEO data migrated to JSON first.

**Comment by SCdF** [comment]
@michaelkohn so to be clear, you're currently parsing XML using SQL and using those values in analytics?

**Comment by michaelkohn** [comment]
@SCdF Not for dashboards, but we are working on a number of other data projects where I'm parsing XML using SQL to get GEO data.  For example... one of the projects we just started working on combines GEO data generated by CHWs with high res settlement imagery from another partner to try to identify households that are not being visited by CHWs (but should).

**Comment by SCdF** [comment]
@michaelkohn wow that's horrifying. OK that definitely blocks us from running the migration, and definitely raises the priority on fixing medic-couch2pg so it does this for you, or at least so that it also downloads attachments and stores them in an `attachments` table or something.

**Comment by SCdF** [comment]
tl;dr; view generation speed is doubled.

---

So I've done some performance tests on this. I did this on my MBP, using a PROD dataset of documents under one branch manager that I used awhile ago to test phone performance / debug crashing. I got everything setup, and then ran the following code in the chrome debug console:
```js
muda = new PouchDB('medic-user-da'); muda.viewCleanup().then(() => {start = Date.now(); return muda.query('medic-client/people_by_phone');}).then(result => console.log(`[${result.total_rows}] ${Date.now() - start}ms`));
```

The view cleanup was superstitious and didn't really help, but I was attempting to remove some extra caching I was seeing and just left it in there. I ran this code 5 times for the normal data set, and then 5 times after running the migration script. I changed the view each time by adding on some no-op code into the view:
```js
function(doc) {
  // Each test had a different if statement
  if (doc.type === 'person' {
  if (doc.type === 'person' || false) {
  if (doc.type === 'person' || false || false) {
  if (doc.type === 'person' && true) {
  if (doc.type === 'person' && true && true) {
    emit([doc.phone]);
  }
}
```

This changed forced a app refresh and the view to regenerate. Numbers in millis:

Pre: 39485, (|f) 39947, (|f|f) 42539, (&t) 40993, (&t&t) 41026
Post: 14416, (|f) [1], (|f|f) 15044, (&t) 15472 (&ttT) 12208

So that's a 2-3x speedup in view generation.

In terms of initial clean load there was some difference (42min vs 36 min), but on such a large dataset on an actively used computer without the interest to run it multiple times I don't think it's a useful metric.

Memory usage was also kind of hard to measure, but within the initial load it seemed to be lower (1.8GB max vs. 1.4GB) but again it's hard to tell.

[1] we lost this one due to complications in Chrome and me thinking it had crashed. You only get one chance to run it unfortunately before after that it's cached. Also after this I ran a PouchDB compaction, which took forever but didn't seem to affect view generation speeds.

**Comment by SCdF** [comment]
View generation is the main benefit of this: I'm not sure what else we'd want to test. I don't think general app usage would be faster per se, because that's all done on views and once we use them they're already generated. However, if the views are stale this speed up will help it catch up.

**Comment by SCdF** [comment]
Here is the plan:
 - Write or formalise an existing script that wraps migration-style scripts so there is a standard way to run them manually
 - Commit this migration script to master, but not to the automatic migrations directly, to somewhere else where we can store migrations that we may want in the future but not to run automatically

**Comment by SCdF** [comment]
@mandric does a script exist to wrap migration-style scripts?

**Comment by SCdF** [comment]
@garethbowen PR plz: https://github.com/medic/medic-api/pull/126

**Comment by garethbowen** [comment]
Comments on the commits.

**Comment by SCdF** [comment]
@garethbowen back to you
